pkg/fqdn/service/service_test.go

```diff
@@ -0,0 +1,260 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/netip"
+	"testing"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/fqdn/namemanager"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/testutils"
+	testidentity "github.com/cilium/cilium/pkg/testutils/identity"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+func TestFQDNDataServer(t *testing.T) {
+
+	test := map[string]struct {
+		port                     int
+		enableL7Proxy            bool
+		enableStandaloneDNSProxy bool
+		serverPort               int
+		err                      error
+	}{
+		"Successfully running the server": {
+			port:                     1234,
+			enableL7Proxy:            true,
+			enableStandaloneDNSProxy: true,
+			serverPort:               40045,
+			err:                      nil,
+		},
+		"Failure on running the server due to invalid server port": {
+			port:                     1234,
+			serverPort:               0,
+			enableStandaloneDNSProxy: true,
+			enableL7Proxy:            true,
+			err:                      errors.New("listen tcp: address -1: invalid port"),
```

Comment:
isn't zero a valid port to listen on? where do you check for this error?

Reply:
So this is one of the limitations as of now for SDP. Since the SDP(client side) needs to be aware of the port on which the grpc server is serving. If we have a port as zero and randomly assign a port that is unused, SDP has no way of currently knowing which port to connect to. Right this same port will be defined in the SDP config map.

I think this can be improved but for now, there is a check in the cell.go: https://github.com/cilium/cilium/pull/38669/files#diff-af3666b3fad42e3a42b2eb689adaa76efa1b185a9640d80f04fdbcb78161b46aR60 that has the check.

Reply:
I see. But that's a different error than `"listen tcp: address -1: invalid port"`; you never seem to check for that specific error anywhere?

Reply:
I realized not returning that error, just logging it. The client should see the error message as the server won't be listening on it. 
If there is way we can check the logs from the cilium agent perspective in the test, then that would be helpful as well.